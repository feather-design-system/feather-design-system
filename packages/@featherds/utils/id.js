const getSafeId = function (str) {
  str = str || "feather";
  const random = Math.floor(Math.random() * 1000000000);
  return [str.replace(/\s+/g, "-"), Date.now(), random].join("-");
};

export { getSafeId };
