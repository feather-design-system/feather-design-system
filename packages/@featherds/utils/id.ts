const getSafeId = function (str: string) {
  str = str || "feather";
  const random = Math.floor(Math.random() * 1000000000);
  return [str.replace(/\s+/g, "-"), Date.now(), random].join("-");
};

export { getSafeId };
