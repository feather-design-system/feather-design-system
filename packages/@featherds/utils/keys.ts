const KEYCODES = {
  PAGEUP: 33,
  PAGEDOWN: 34,
  UP: 38,
  DOWN: 40,
  LEFT: 37,
  RIGHT: 39,
  HOME: 36,
  END: 35,
  ENTER: 13,
  SPACE: 32,
  ESCAPE: 27,
  BACKSPACE: 8,
  DELETE: 46,
};

export { KEYCODES };
