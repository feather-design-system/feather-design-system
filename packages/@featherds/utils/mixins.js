var translateKey = {
  props: {
    translationKey: {
      type: String,
      default: "",
    },
  },
  methods: {
    translate: function (key) {
      return this.$t
        ? this.$t(this.translationKey + key)
        : "Vue i18n is not loaded";
    },
  },
};
export { translateKey };
