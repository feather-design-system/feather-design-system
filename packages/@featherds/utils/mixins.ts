const translateKey = {
  props: {
    translationKey: {
      type: String,
      default: "",
    },
  },
  methods: {
    translate: function (this: IComponent, key: string) {
      return this.$t
        ? this.$t(this.translationKey + key)
        : "Vue i18n is not loaded";
    },
  },
};

interface IComponent {
  translationKey: string;
  $t: (s: string) => void;
}
export { translateKey };
