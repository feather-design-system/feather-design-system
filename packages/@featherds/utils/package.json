{
  "name": "@featherds/utils",
  "version": "0.2.10",
  "publishConfig": {
    "access": "public"
  },
  "author": "NantHealth",
  "scripts": {},
  "keywords": [
    "vue",
    "FeatherUtils"
  ],
  "types": "./src/index.d.ts"
}
