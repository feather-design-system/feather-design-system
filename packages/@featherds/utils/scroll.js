var toView = function (el, parent) {
  if (!el || !parent) {
    return;
  }
  var elHeight = el.getBoundingClientRect().height;
  const compStyle = getComputedStyle(el);
  elHeight += parseInt(compStyle["margin-top"], 10);
  elHeight += parseInt(compStyle["margin-bottom"], 10);
  parent.scrollTop =
    el.offsetTop - parent.getBoundingClientRect().height + elHeight;
};

export { toView };
