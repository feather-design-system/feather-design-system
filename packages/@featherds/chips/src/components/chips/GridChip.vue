<template>
  <Chip
    v-bind="attrs"
    :disabled="disabled"
    :condensed="condensed"
    :class="{ hover: canClick, focus: canClick || canDelete }"
    role="row"
    :clickable="canClick"
  >
    <span role="gridcell" :aria-disabled="disabled">
      <span v-bind="chipTextAttrs" class="chip-label-button"
        ><PreIcon v-if="hasIcon"><slot name="icon" /></PreIcon>
        <Label :id="chipTextId"><slot /></Label
      ></span>
    </span>
    <DeleteIcon
      v-if="canDelete"
      :disabled="disabled"
      :text-id="chipTextId"
      :label="deleteLabel"
      role="gridcell"
      @delete="$emit('delete')"
    />
  </Chip>
</template>
<script lang="ts">
import { getSafeId } from "@featherds/utils/id";
import { useLabelProperty } from "@featherds/composables/LabelProperty";
import { toRef, computed, defineComponent, PropType } from "vue";
import Chip from "../base/Chip.vue";
import DeleteIcon from "../base/DeleteIcon.vue";
import Label from "../base/Label.vue";
import PreIcon from "../base/PreIcon.vue";

const LABELS = {
  delete: "Remove",
};
export default defineComponent({
  inheritAttrs: false,
  props: {
    disabled: {
      type: Boolean,
      default: false,
    },
    labels: {
      type: Object as PropType<typeof LABELS>,
      default: () => {
        return LABELS;
      },
    },
    condensed: {
      type: Boolean,
      default: false,
    },
  },
  setup(props, context) {
    const labels = useLabelProperty(toRef(props, "labels"), LABELS);
    const chipTextId = computed(() => getSafeId("chip-text"));

    const handleClick = () => {
      if (props.disabled) {
        return;
      }
      context.emit("click");
    };

    const _attrs = { ...context.attrs };

    if (props.disabled) {
      delete _attrs.onClick;
    }

    return {
      ...labels,
      chipTextId,
      handleClick,
      attrs: _attrs,
    };
  },
  computed: {
    chipTextAttrs() {
      if (this.canClick) {
        return {
          role: "button",
          tabindex: "0",
        };
      }
      return {};
    },
    canDelete() {
      return !!this.$attrs.onDelete && !this.disabled;
    },
    canClick() {
      return !!this.$attrs.onClick && !this.disabled;
    },
    hasIcon() {
      const hasIcon =
        this.$slots.icon &&
        this.$slots
          .icon()
          .findIndex(
            (o) =>
              (o.children && o.children.length !== 0) ||
              typeof o.type === "object"
          ) !== -1;
      return hasIcon;
    },
  },
  components: {
    Chip,
    DeleteIcon,
    Label,
    PreIcon,
  },
});
</script>

<style lang="scss" scoped>
@import "@featherds/styles/themes/variables";
@import "@featherds/styles/mixins/typography";
@import "../../../scss/mixins";
[role="gridcell"] {
  height: 100%;
}
.chip-label-button {
  height: 100%;
  display: inline-flex;
  align-items: center;
  &:focus {
    outline: 0;
  }
}
</style>
