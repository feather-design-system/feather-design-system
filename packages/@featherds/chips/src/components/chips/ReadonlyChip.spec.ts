import GridChip from "./GridChip.vue";
import { mount } from "@vue/test-utils";
import * as id from "@featherds/utils/id";
jest.spyOn(id, "getSafeId").mockImplementation((x) => x);

const getWrapper = function (options = {}, addIcon = false) {
  options = {
    ...options,
    ...getSlots(addIcon),
  };

  return mount(GridChip, options);
};
const getSlots = (withIcon = false) => {
  const slots = {
    slots: {
      default: "Test",
    },
  } as {
    slots: {
      default: string;
      icon?: { template: string };
    };
  };
  if (withIcon) {
    slots.slots.icon = { template: "<span>*</span>" };
  }
  return slots;
};

describe("ReadOnlyChip", () => {
  it("should render a standard chip", () => {
    const wrapper = getWrapper();
    expect(wrapper.element).toMatchSnapshot();
  });
  it("should render a standard chip with icon", () => {
    const wrapper = getWrapper({}, true);
    expect(wrapper.element).toMatchSnapshot();
  });
  it("should render a disabled chip", () => {
    const props = {
      disabled: true,
    };
    const wrapper = getWrapper({ props });
    expect(wrapper.element).toMatchSnapshot();
  });

  it("should render condensed", () => {
    const props = {
      condensed: true,
    };
    const wrapper = getWrapper({ props });
    expect(wrapper.element).toMatchSnapshot();
  });
  it("should render condensed from a group", () => {
    const provide = {
      condensed: true,
    };
    const wrapper = getWrapper({ global: { provide: provide } });
    expect(wrapper.element).toMatchSnapshot();
  });
});
