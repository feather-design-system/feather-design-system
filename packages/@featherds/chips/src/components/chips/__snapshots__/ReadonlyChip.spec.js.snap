// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReadOnlyChip should render a disabled chip 1`] = `
<div
  aria-disabled="true"
  class="chip disabled"
  role="row"
>
  
  <span
    aria-disabled="true"
    role="gridcell"
  >
    <span
      class="chip-label-button"
    >
      <!--v-if-->
      <span
        class="label"
        id="chip-text"
        title="false"
      >
        
        
        Test
        
        
      </span>
    </span>
  </span>
  <!--v-if-->
  
  <!--v-if-->
</div>
`;

exports[`ReadOnlyChip should render a standard chip 1`] = `
<div
  aria-disabled="false"
  class="chip"
  role="row"
>
  
  <span
    aria-disabled="false"
    role="gridcell"
  >
    <span
      class="chip-label-button"
    >
      <!--v-if-->
      <span
        class="label"
        id="chip-text"
        title="false"
      >
        
        
        Test
        
        
      </span>
    </span>
  </span>
  <!--v-if-->
  
  <!--v-if-->
</div>
`;

exports[`ReadOnlyChip should render a standard chip with icon 1`] = `
<div
  aria-disabled="false"
  class="chip"
  role="row"
>
  
  <span
    aria-disabled="false"
    role="gridcell"
  >
    <span
      class="chip-label-button"
    >
      <span
        class="chip-icon"
        role="presentation"
      >
        
        
        <span>
          *
        </span>
        
        
      </span>
      <span
        class="label"
        id="chip-text"
        title="false"
      >
        
        
        Test
        
        
      </span>
    </span>
  </span>
  <!--v-if-->
  
  <!--v-if-->
</div>
`;

exports[`ReadOnlyChip should render condensed 1`] = `
<div
  aria-disabled="false"
  class="chip condensed"
  role="row"
>
  
  <span
    aria-disabled="false"
    role="gridcell"
  >
    <span
      class="chip-label-button"
    >
      <!--v-if-->
      <span
        class="label"
        id="chip-text"
        title="false"
      >
        
        
        Test
        
        
      </span>
    </span>
  </span>
  <!--v-if-->
  
  <!--v-if-->
</div>
`;

exports[`ReadOnlyChip should render condensed from a group 1`] = `
<div
  aria-disabled="false"
  class="chip"
  role="row"
>
  
  <span
    aria-disabled="false"
    role="gridcell"
  >
    <span
      class="chip-label-button"
    >
      <!--v-if-->
      <span
        class="label"
        id="chip-text"
        title="false"
      >
        
        
        Test
        
        
      </span>
    </span>
  </span>
  <!--v-if-->
  
  <!--v-if-->
</div>
`;
